tools/pipeline-witness/Azure.Sdk.Tools.PipelineWitness/BlobUploadProcessor.cs

```diff
@@ -116,17 +159,106 @@ public BlobUploadProcessor(ILogger<BlobUploadProcessor> logger, BuildLogProvider
+                }
+            }            
+        }
+
+        public async Task ProcessBuildLogBundleAsync(BuildLogBundle buildLogBundle)
+        {
+            foreach (var log in buildLogBundle.TimelineLogs)
+            {
+                await UploadLogLinesBlobAsync(buildLogBundle, log);
+            }
+        }
+
+        private List<BuildLogBundle> BuildLogBundles(string account, Build build, Timeline timeline, List<BuildLog> logs)
+        {
+            BuildLogBundle CreateBundle() => new BuildLogBundle
+            {
+                Account = account,
+                BuildId = build.Id,
+                ProjectId = build.Project.Id,
+                ProjectName = build.Project.Name,
+                QueueTime = build.QueueTime.Value,
+                StartTime = build.StartTime.Value,
+                FinishTime = build.FinishTime.Value,
+                DefinitionId = build.Definition.Id,
+                DefinitionName = build.Definition.Name,
+                DefinitionPath = build.Definition.Path
+            };
+
+            BuildLogBundle currentBundle;
+            var logBundles = new List<BuildLogBundle>();
+            logBundles.Add(currentBundle = CreateBundle());
+
+            var logsById = logs.ToDictionary(l => l.Id);
+
+            foreach (var log in logs)
+            {
+                if(currentBundle.TimelineLogs.Count >= this.options.Value.BuildLogBundleSize)
+                {
+                    logBundles.Add(currentBundle = CreateBundle());
+                }
+
+                var logRecords = timeline.Records.Where(x => x.Log?.Id == log.Id).ToArray();
+
+                if(logRecords.Length > 1)
+                {
+                    this.logger.LogWarning("Found multiple timeline records for build {BuildId}, log {LogId}", build.Id, log.Id);
+                }
+
+                var logRecord = logRecords.FirstOrDefault();
+
+                // Job logs are typically just a duplication of their child task logs with the addition of extra start and end lines.
+                // If we can, we skip the redundant lines.
+                if (string.Equals(logRecord?.RecordType, "job", StringComparison.OrdinalIgnoreCase))
                 {
-                    await UploadLogLinesBlobAsync(account, build, log);
+                    // find all of the child task records
+                    var childRecords = timeline.Records.Where(x => x.ParentId == logRecord.Id);
+                    
+                    // sum the line counts for all of the child task records
+                    var childLineCount = childRecords
+                        .Where(x => x.Log != null && logsById.ContainsKey(x.Log.Id))
+                        .Sum(x => logsById[x.Log.Id].LineCount);
+                    
+                    // if the job's line count is the task line count + 2, then we can skip the job log
+                    if (log.LineCount == childLineCount + 2)                    
```

Comment:
Are there other cases where the extra lines are not just empty padding? If not, could this code be simplified to trim the empty start and end lines and then compare the hashes of the two strings?

Reply:
I'm also curious whether this is slow for larger logs or not, if I'm understanding that this code will search all log lines from a job for a log id.

Reply:
> Are there other cases where the extra lines are not just empty padding?

The extra lines aren't just empty padding.  They're timestamped begin / end traces:

Job:
https://dev.azure.com/azure-sdk/public/_apis/build/builds/1489836/logs/135
```
2022-04-07T16:07:34.5175923Z ##[section]Starting: Compliance
2022-04-07T16:07:34.6907825Z ##[section]Starting: Initialize job
2022-04-07T16:07:34.6909373Z Agent name: 'azsdk-pool-mms-win-2019-general 3'
2022-04-07T16:07:34.6909739Z Agent machine name: 'a83a2547c000001'
2022-04-07T16:07:34.6909931Z Current agent version: '2.202.0'
```

1st Task:
https://dev.azure.com/azure-sdk/public/_apis/build/builds/1489836/logs/7

```
2022-04-07T16:07:34.6907728Z ##[section]Starting: Initialize job
2022-04-07T16:07:34.6909351Z Agent name: 'azsdk-pool-mms-win-2019-general 3'
2022-04-07T16:07:34.6909733Z Agent machine name: 'a83a2547c000001'
2022-04-07T16:07:34.6909926Z Current agent version: '2.202.0'
```

Interleaved by timestamp
```
Job  2022-04-07T16:07:34.5175923Z ##[section]Starting: Compliance
Task 2022-04-07T16:07:34.6907728Z ##[section]Starting: Initialize job
Job  2022-04-07T16:07:34.6907825Z ##[section]Starting: Initialize job
Task 2022-04-07T16:07:34.6909351Z Agent name: 'azsdk-pool-mms-win-2019-general 3'
Job  2022-04-07T16:07:34.6909373Z Agent name: 'azsdk-pool-mms-win-2019-general 3'
Task 2022-04-07T16:07:34.6909733Z Agent machine name: 'a83a2547c000001'
Job  2022-04-07T16:07:34.6909739Z Agent machine name: 'a83a2547c000001'
Task 2022-04-07T16:07:34.6909926Z Current agent version: '2.202.0'
Job  2022-04-07T16:07:34.6909931Z Current agent version: '2.202.0'
```

Looking back at the last 5 months of data, there were no instances of Jobs containing 2 more lines than there child tasks where the Job's logs weren't all duplicated task logs.

>  If not, could this code be simplified to trim the empty start and end lines and then compare the hashes of the two strings?

We could be more exact and keep a line number / message hash pair for every task line, then skip job lines if their line number / has is already in the set,  We're intentionally avoiding every having even a single full task log in memory as some tasks are known to produce very log lines.  We may need to do even more here as our stream reader over http response stream may keep more in memory than we need for a line-at-a-time process.

> I'm also curious whether this is slow for larger logs or not, if I'm understanding that this code will search all log lines from a job for a log id.

This pulls the build logs metadata, not the log contents. It's a single rest call to get all the log ids and their line counts.  We compare that to the TimeLine data that has all the timeline records (task/job/stage) with their associated log ids. This is all compute against data in-hand with a natural set size limiter of the number of tasks that can appear in a build timeline.   At most, we'll be scanning a set of maybe thousands of task objects for a few hundred job ids.